client/securedrop_client/sdk/__init__.py

```diff
@@ -147,12 +164,102 @@ def _send_json_request(
         if timeout:
             data["timeout"] = timeout
 
+        env = {}
+        if self.development_mode:
+            env["SD_PROXY_ORIGIN"] = self.server
+
+        # Streaming
+        if stream:
+            fobj = tempfile.TemporaryFile("w+b")
+
+            content_length = 0
+            filename = ""
+            sha256sum = ""
+
+            retry = 0
+            bytes_written = 0
+            download_finished = False
+
+            while not download_finished and retry < DOWNLOAD_RETRY_LIMIT:
+                try:
+                    # Open the process
+                    proc = subprocess.Popen(
+                        self._rpc_target(),
+                        stdin=subprocess.PIPE,
+                        stdout=subprocess.PIPE,
+                        stderr=subprocess.PIPE,
+                        env=env,
+                    )
+
+                    # Send the data, included the range if we are retrying
+                    if retry > 0:
+                        data["headers"]["Range"] = f"bytes={bytes_written}-"
+
+                    data_str = json.dumps(data).encode()
+                    if proc.stdin is not None:
+                        proc.stdin.write(data_str)
+                        proc.stdin.close()
+
+                    # Save the response headers from the first request
+                    if retry == 0:
+                        try:
+                            if proc.stderr is not None:
+                                stderr = json.loads(proc.stderr.read().decode())
+                                sha256sum = stderr["headers"]["etag"]
+                                filename = stderr["headers"]["content-disposition"]
+                                content_length = int(stderr["headers"]["content-length"])
+                        except (json.decoder.JSONDecodeError, KeyError) as err:
+                            raise BaseError(
+                                "Unable to parse header metadata from response"
+                            ) from err
+
+                    # Write the contents to disk
+                    chunk_size = 64
+                    while not download_finished:
+                        if proc.stdout is not None:
+                            chunk = proc.stdout.read(chunk_size)
+                            if not chunk:
+                                download_finished = True
+                                break
+
+                        fobj.write(chunk)
+                        bytes_written += len(chunk)
+                        if bytes_written == content_length:
+                            download_finished = True
+
+                    # FIXME: For now, store the contents as bytes
+                    fobj.seek(0)
+                    contents = fobj.read()
```

Comment:
If there is an HTTP error (any 4XX or 5XX status code), then the proxy will return a JSONResponse, which is why the previous code checked `response.stdout[0] != b"{"` to see if it was a JSON blob